extensions/wrapper/clients/typescript-client/src/oauth2/AccessTokenService.ts

```diff
@@ -0,0 +1,54 @@
+import {ClientCredentialsResponse, fetchClientCredentials} from './FetchUtils';
+import {buildTokenRequestFormData} from './RequestUtils';
+import {ClientCredentials} from './model/ClientCredentials';
+
+export class AccessTokenService {
+    private activeRequest: Promise<ClientCredentialsResponse> | null = null;
+    private refreshTimeout?: NodeJS.Timeout;
+    private tokenRequestFormData: string;
+    private accessToken: string | null = null;
+
+    constructor(private clientCredentials: ClientCredentials) {
+        this.tokenRequestFormData = buildTokenRequestFormData(
+            clientCredentials.clientId,
+            clientCredentials.clientSecret,
+        );
```

Comment:
However, "form data" is a concern for the fetch request and should not be in this service, that wants to delegate all concerns to the fetch method, so it can focus on "when to fetch, what to store". it is fine to have a function: `fetchAccessToken(credentials: ClientCredentials): ClientCredentialsResponse`